fhir-persistence-cos/src/main/java/com/ibm/fhir/persistence/cos/client/COSPayloadClient.java

```diff
@@ -0,0 +1,232 @@
+     * @param bundleName
+     * @param compressedPayload the serialized payload (already compressed)
+     */
+    public void write(String objectName, byte[] compressedPayload) throws FHIRPersistenceException {
+        final String bucketName = getBucketName();
+
+        // Set up the metadata for the call to S3/COS
+        ObjectMetadata omd = new ObjectMetadata();
+        omd.setContentLength(compressedPayload.length);
+        omd.setContentType("binary");
+
+        if (logger.isLoggable(Level.FINE)) {
+            logger.info("Writing to COS '" + bucketName + ":" + objectName + "', bytes: " + compressedPayload.length);
+        }
+
+        // Write the object to the target key (objectName) in the given bucket
+        InputStream inputStream = new ByteArrayInputStream(compressedPayload);
+        PutObjectResult result = client.putObject(new PutObjectRequest(bucketName, objectName, inputStream, omd));
+        if (result != null) {
+            if (logger.isLoggable(Level.FINE)) {
+                logger.fine("Wrote [" + bucketName + "]/" + objectName + ", ETag: " + result.getETag());
+            }
+        } else {
+            logger.warning("Writing failed for [" + bucketName + "]/" + objectName + ", bytes: " + compressedPayload.length);
+            throw new RuntimeException("Write to COS failed");
+        }
+    }
+
+    /**
+     * Write the payload held in the ioStream object to the given objectName as key
+     * @param objectName
+     * @param ioStream
+     * @throws FHIRPersistenceException
+     */
+    public void write(String objectName, InputOutputByteStream ioStream) throws FHIRPersistenceException {
+        final String bucketName = getBucketName();
+
+        // Set up the metadata for the call to S3/COS
+        ObjectMetadata omd = new ObjectMetadata();
+        omd.setContentLength(ioStream.size());
+        omd.setContentType("binary");
+
+        if (logger.isLoggable(Level.FINE)) {
+            logger.info("Writing to COS '" + bucketName + ":" + objectName + "', bytes: " + ioStream.size());
+        }
+
+        // Write the object to the target key (objectName) in the given bucket
+        PutObjectResult result = client.putObject(new PutObjectRequest(bucketName, objectName, ioStream.inputStream(), omd));
+        if (result != null) {
+            if (logger.isLoggable(Level.FINE)) {
+                logger.fine("Wrote [" + bucketName + "]/" + objectName + ", ETag: " + result.getETag());
+            }
+        } else {
+            logger.warning("Writing failed for [" + bucketName + "]/" + objectName + ", bytes: " + ioStream.size());
+            throw new RuntimeException("Write to COS failed");
+        }
+    }
+
+    /**
+     * Create the bucket if it doesn't yet exist
+     * @param bucketName
+     */
+    public void createBucketIfNeeded(String bucketName) throws FHIRPersistenceException {
+        if (!client.doesBucketExistV2(bucketName)) {
```

Comment:
This is messy... bucketname can exist yet the api returns false when using virtual hosts.